src/main/java/com/metamx/common/FileUtils.java

```diff
@@ -0,0 +1,302 @@
+   *
+   * @throws IOException passed from inner stream errors
+   */
+  public static Boolean checkDigest(final File file, final MessageDigest otherDigest) throws IOException
+  {
+    return FileCopyResult.buildResult(file.toPath(), otherDigest.getAlgorithm()).digestEquals(otherDigest);
+  }
+
+  /**
+   * Copy input stream to outFile. If outFile exists, it is attempted to be deleted. Also validates the outFile with MD5
+   *
+   * @param inputStreamSupplier Supplier for an input stream that is to be copied. The resulting stream is closed each iteration
+   * @param shouldRetry         Predicate indicating if an error is recoverable and should be retried. Note that MD5MismatchException may be thrown by the inner loop in addition to IOException
+   * @param outFile             Where the file should be written to.
+   * @param maxAttempts         The maximum number of assumed recoverable attempts to try before completely failing.
+   *
+   * @throws java.lang.RuntimeException wrapping the inner exception on failure.
+   */
+  public static FileCopyResult retryCopy(
+      final ByteSource inputStreamSupplier,
+      final Predicate<Throwable> shouldRetry,
+      final File outFile,
+      final int maxAttempts
+  )
+  {
+    try {
+      return RetryUtils.retry(
+          new Callable<FileCopyResult>()
+          {
+            @Override
+            public FileCopyResult call() throws Exception
+            {
+              try (final InputStream inputStream = inputStreamSupplier.openStream()) {
+                if (outFile.exists()) {
+                  if (!outFile.delete()) {
+                    throw new IAE("Could not remove file at [%s]", outFile.getAbsolutePath());
+                  }
+                }
+
+                final MessageDigest digest = StreamUtils.md5StreamToFile(inputStream, outFile);
+                final FileCopyResult result = FileCopyResult.buildResult(outFile.toPath(), "MD5");
+                if (!result.digestEquals(digest)) {
+                  throw new MD5MismatchException(
+                      "File at [%s] did not match MD5 after copy",
+                      outFile.getAbsolutePath()
+                  );
+                }
+                return result;
+              }
+            }
+          },
+          shouldRetry,
+          maxAttempts
+      );
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Retry copy attempts from input stream to output stream. Does *not* check to make sure data was intact during the transfer
+   *
+   * @param inputStreamSupplier  Supplier for input streams to copy from. The stream is closed on every retry.
+   * @param shouldRetry          Predicate to determine if the throwable is recoverable for a retry
+   * @param outputStreamSupplier Supplier for output streams. The stream is closed on every retry.
+   * @param maxAttempts          Maximum number of retries before failing
+   */
+  public static long retryCopy(
```

Comment:
having two methods with the same name but quite different signatures seems a bit confusing.